### PR TITLE
Feature/dm/rule scope

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: server/sonar-web/src/main/js/components/icons-components/RuleScopeIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      className={className}
+      width={size}
+      height={size}
+      viewBox="0 0 16 16"
+      version="1.1"
+      xmlnsXlink="http://www.w3.org/1999/xlink"
+      xmlSpace="preserve">
+      <path
+        style={{ fill }}
+        d="M8 3.071c2.724 0 4.929 2.204 4.929 4.929s-2.204 4.929-4.929 4.929c-2.724 0-4.929-2.204-4.929-4.929s2.204-4.929 4.929-4.929zM8 1.357c-3.669 0-6.643 2.974-6.643 6.643s2.974 6.643 6.643 6.643 6.643-2.974 6.643-6.643-2.974-6.643-6.643-6.643zM8 6.286c0.945 0 1.714 0.769 1.714 1.714s-0.769 1.714-1.714 1.714-1.714-0.769-1.714-1.714 0.769-1.714 1.714-1.714zM8 4.571c-1.893 0-3.429 1.535-3.429 3.429s1.535 3.429 3.429 3.429 3.429-1.535 3.429-3.429-1.535-3.429-3.429-3.429z"
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/icons-components/RuleScopeIcon.tsx]: Code Smell Code Smell: Exceeds maximum line length of 180 ([typescript:S103](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS103))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fdm%2Frule_scope&issues=AWFHHYDYKuAw5oyFANwr&open=AWFHHYDYKuAw5oyFANwr)

[ISSUE_KEY]: AWFHHYDYKuAw5oyFANwr

File: server/sonar-server/src/main/java/org/sonar/server/rule/RuleCreator.java [hunk 2]
```diff
@@ -39,6 +39,7 @@
 import org.sonar.db.rule.RuleDefinitionDto;
 import org.sonar.db.rule.RuleDto;
 import org.sonar.db.rule.RuleDto.Format;
+import org.sonar.db.rule.RuleDto.Scope;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RuleCreator.java]: Code Smell Code Smell: Remove this unused import 'org.sonar.db.rule.RuleDto.Scope'. ([squid:UselessImportCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fdm%2Frule_scope&issues=AWFHHYE2KuAw5oyFANws&open=AWFHHYE2KuAw5oyFANws)

[ISSUE_KEY]: AWFHHYE2KuAw5oyFANws

File: server/sonar-server/src/main/java/org/sonar/server/rule/ws/SearchAction.java [hunk 3]
```diff
@@ -145,7 +145,8 @@ public void define(WebService.NewController controller) {
         "<li>\"debtRemFn\" becomes \"remFn\"</li>" +
         "<li>\"effortToFixDescription\" becomes \"gapDescription\"</li>" +
         "<li>\"debtOverloaded\" becomes \"remFnOverloaded\"</li>" +
-        "</ul>")
+        "</ul>" +
+        "In 7.1, the field 'scope' has been added.")
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/rule/ws/SearchAction.java]: Please remove this from the description and add it to the changelog of the action (see next comment)

File: server/sonar-server/src/main/java/org/sonar/server/rule/ws/SearchAction.java [hunk 4]
```diff
@@ -190,7 +191,8 @@ private void doDefinition(WebService.NewAction action) {
       "<li>\"debtRemFnOffset\" becomes \"remFnBaseEffort\"</li>" +
       "<li>\"defaultDebtRemFnOffset\" becomes \"defaultRemFnBaseEffort\"</li>" +
       "<li>\"debtOverloaded\" becomes \"remFnOverloaded\"</li>" +
-      "</ul>")
+      "</ul>" +
+      "In 7.1, a new field 'scope' has been added to the response.")
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/rule/ws/SearchAction.java]: Please move this line to the changelog of the action : 

```
action.setDescription("Search for a collection...")
   .setChangelog(new Change("7.1", "new field 'scope' has been added to the response")
   .setChangelog(new Change("7.1", "field 'scope' has been added to 'f' parameter")
```

File: sonar-plugin-api/src/main/java/org/sonar/api/rule/RuleScope.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.api.rule;
+
+/**
+ * @since 7.1
+ *
+ */
+public enum RuleScope {
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/rule/RuleScope.java]: I would tend to put this class as an inner class of RulesDefinition, as it should only be used with RulesDefinition

File: sonar-core/src/main/resources/org/sonar/l10n/core.properties [hunk 6]
```diff
@@ -1302,6 +1302,11 @@ coding_rules.remediation_function.coeff=Coeff
 coding_rules.remediation_function.offset=Offset
 coding_rules.remediation_function.constant=Constant
 
+coding_rules.scope.title=Analysis Scope
+coding_rules.scope.MAIN=Main sources
+coding_rules.scope.TESTS=Test sources
```
Comment (julienlancelot) [inline sonar-core/src/main/resources/org/sonar/l10n/core.properties]: Typo in scope TESTS, it should be TEST => 

`coding_rules.scope.TEST=Test sources`

Comment (julienlancelot) [reply sonar-core/src/main/resources/org/sonar/l10n/core.properties]: (The UI is badly displaying label coding_rules.scope.TESTS when a rule has a TEST scope)

File: server/sonar-web/src/main/js/components/icons-components/RuleScopeIcon.tsx (hunk 1 above)
Comment (SonarTech) [reply server/sonar-web/src/main/js/components/icons-components/RuleScopeIcon.tsx]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: sonar-core/src/main/resources/org/sonar/l10n/core.properties (hunk 6 above)
Comment (gregaubert) [reply sonar-core/src/main/resources/org/sonar/l10n/core.properties]: Good catch! Thanks!

File: pom.xml [hunk 7]
```diff
@@ -49,16 +49,16 @@
   <properties>
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply pom.xml]: Must be a mistake with git, good catch

File: server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v71/SetRuleScopeToMainTest.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    assertThat(countRowsWithValue(RuleScope.MAIN)).isEqualTo(0);
+
+    underTest.execute();
+
+    assertThat(countRowsWithValue(null)).isEqualTo(0);
+    assertThat(countRowsWithValue(RuleScope.MAIN)).isEqualTo(2);
+  }
+
+  @Test
+  public void support_large_number_of_rows() throws SQLException {
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v71/SetRuleScopeToMainTest.java]: Indeed, I had copied the tests from a similar migration.

File: server/sonar-server/src/main/java/org/sonar/server/rule/RuleCreator.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RuleCreator.java]: Code Smell Code Smell: Remove this unused import 'org.sonar.db.rule.RuleDto.Scope'. ([squid:UselessImportCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fdm%2Frule_scope&issues=AWFHHYE2KuAw5oyFANws&open=AWFHHYE2KuAw5oyFANws)

[ISSUE_KEY]: AWFHHYE2KuAw5oyFANws

**Pull request merged**